### PR TITLE
[WIP] generate bindings using libclang

### PR DESCRIPTION
**Pull request by jonas**
This is still early stage, but it would be great to have some feedback regarding what to focus on at this point and ease the way to have it integrated. For now it is focused on internal usage and a good use case will be to help complete #104. However, for this to happen there are some things to figure out such as how to sort the read C types to make them stable across systems, whether to use white- or black-listing to "weed" out non-POSIX/standard types etc.

Some initial thoughts can be found in [docs/contrib/bindgen.rst](https://github.com/jonas/scala-native/blob/bindgen/docs/contrib/bindgen.rst)
And a dummy test parsing [unit-tests/src/main/resources/bindgen/test.h](https://github.com/jonas/scala-native/blob/bindgen/unit-tests/src/main/resources/bindgen/test.h) is in the [BindgenSuite](https://github.com/jonas/scala-native/blob/bindgen/unit-tests/src/main/scala/scala/scalanative/runtime/bindgen/BindgenSuite.scala).

References: #154
CC: @frgomes

### CODE REVIEW

File: nativelib/src/main/resources/clang.c [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+#ifdef SCALA
+// Auto-generated from nativelib/src/main/resources/clang.c - DO NOT EDIT
+package scala.scalanative
+package runtime
+package bindgen
+
+import scalanative.native._
+
+#define EXTERN_FUNCTION0(symbol, id, tpe) @name(#symbol) def id(): tpe = extern;
```
Comment (jonas) [inline nativelib/src/main/resources/clang.c]: A new type of Scala macro ;)

File: nativelib/src/main/resources/clang.c [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+  void *data;
+};
+
+static enum CXChildVisitResult
+scalanative_clang_visit(CXCursor cursor, CXCursor parent, CXClientData data)
+{
+  struct bindgen_context *ctx = data;
+
+  CXSourceLocation location = clang_getCursorLocation(cursor);
+  if (!clang_Location_isFromMainFile(location))
```
Comment (jonas) [inline nativelib/src/main/resources/clang.c]: FIXME: Move to Scala code

File: unit-tests/src/main/scala/scala/scalanative/runtime/bindgen/BindgenSuite.scala [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+                               "scala.scalanative.test")
+    val expected =
+      """package scala.scalanative.test
+        |
+        |import scala.scalanative.native._
+        |
+        |@extern
+        |object Test {
+        |  object color {
+        |    val RED = 1
```
Comment (MasseGuillaume) [inline unit-tests/src/main/scala/scala/scalanative/runtime/bindgen/BindgenSuite.scala]: maybe `@inline final val` ?

File: nativelib/src/main/scala/scala/scalanative/runtime/bindgen/Clang.scala [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  def CXType_VariableArray(): CXTypeKind = extern;
+  @name("scalanative_clang_CXType_DependentSizedArray")
+  def CXType_DependentSizedArray(): CXTypeKind = extern;
+  @name("scalanative_clang_CXType_MemberPointer")
+  def CXType_MemberPointer(): CXTypeKind = extern;
+  @name("scalanative_clang_CXType_Auto") def CXType_Auto(): CXTypeKind =
+    extern;
+
+  @name("scalanative_clang_getCursorKind")
+  def getCursorKind(`*cursor`: CXCursor): CXCursorKind = extern;
```
Comment (MasseGuillaume) [inline nativelib/src/main/scala/scala/scalanative/runtime/bindgen/Clang.scala]: hum this `*cursor` looks suspicious.

File: docs/contrib/bindgen.rst [hunk 5]
```diff
@@ -0,0 +1,40 @@
+Automatic binding generation
+============================
+
+Binding generation for now is experimental. The goal is to first use it
+internally to generate bindings for C's stdlib and later make it
+available via scala.meta macros and sbt tasks.
```
Comment (MasseGuillaume) [inline docs/contrib/bindgen.rst]: I would vote for code generation here. The less powerful technique that works well with IDEs

**Comment by MasseGuillaume** [comment]
Really cool! Nice work. I can help on the sbt front ;-).

File: nativelib/src/main/scala/scala/scalanative/runtime/bindgen/Clang.scala (hunk 4 above)
Comment (jonas) [inline nativelib/src/main/scala/scala/scalanative/runtime/bindgen/Clang.scala]: Yes, this is a relic of using C preprocessor to generate the Scala code.

**Comment by jonas** [comment]
The main problem right now is that xcode includes the libclang dynlib but not the libclang header files.

**Comment by jonas** [comment]
Achievement unlocked 🎉 CI passes

File: .travis.yml [hunk 6]
```diff
@@ -46,6 +46,8 @@ matrix:
         - brew link bdw-gc
         - brew install jq
         - brew install re2
+        - brew install llvm@3.7
+      env: PATH="$PATH:/usr/local/opt/llvm@3.7/bin"
```
Comment (MasseGuillaume) [inline .travis.yml]: why 3.7 ?

File: docs/contrib/bindgen.rst (hunk 5 above)
Comment (densh) [inline docs/contrib/bindgen.rst]: IDE support is a blocker here, if scala.meta macros don't work well enough we might have to use code generation instead.

**Comment by densh** [comment]
Amazing progress @jonas! I'm going to have a deeper look at the code itself as soon as 0.2 is shipped.

File: .travis.yml (hunk 6 above)
Comment (jonas) [inline .travis.yml]: To keep the macOS build on the same version, however, we could also change it so CI builds for oldest and newest supported LLVM version.

**Comment by jacob-carlborg** [comment]
> However, for this to happen there are some things to figure out such as how to sort the read C types to make them stable across systems

This is an interesting problem. I work on a binding tool for D [1] myself using libclang as well. When header files start to use `#ifdef` to check for different platforms it starts to get complicated. For example:

```c
#ifdef _WIN32
#include <windows.h>
DWORD foo();
#else
int foo();
#endif
```

The above code would generate different bindings on different platforms. Trying to generate bindings for the Windows branch on a Posix platform wouldn't work because `windows.h` wouldn't be available.

I would also say it depends on how the bindings are used/generated. If the bindings are to be generated with an external tool and placed in a source repository you will have the problem mentioned above. But, if the bindings are generated on the fly by the compiler or a build tool the above problem doesn't matter since you can only interface with the API that is provided on the current platform anyway, making platform dependent bindings not an issue. This is kind of how Swift interfaces with C and Objective-C.

[1] https://github.com/jacob-carlborg/dstep

**Comment by densh** [comment]
It seems like current state of this PR is a bit premature to be included upstream. Especially considering that we're planning to do a major overhaul of the interop layer in 0.4 series. 

Additionally, the binding generation tool can happily leave outside of the main repo in its own subproject and including it into the main repo might aggravate our build time and CI issues we've been having recently. 

I'm closing this PR for now, we can continue discussion on going forward with this on gitter.

**Comment by jonas** [comment]
👍

**Pull request closed without merging**